gh-116126: Implement PEP 696 (#116129)
python/cpython: Doc/library/typing.rst

.. _typevar:

> .. class:: TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False, infer_variance=False, default=None)

Reviewer: I don't think the signature of `TypeVar` itself should have `default=None`, because otherwise it's impossible to distinguish a TypeVar without a default from one that actually defaults to having type `None` (https://github.com/python/typing_extensions/pull/85)

Author: Good point, that's actually currently incorrect in the runtime code: `TypeVar("T")` and `TypeVar("T", default=None)` cannot be distinguished. Do you know how we should present this in the docs? Maybe this:
```suggestion
.. class:: TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False, infer_variance=False, default=<marker>)
```

Reviewer: not sure. Some other options:

```suggestion
.. class:: TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False, infer_variance=False, default=<unrepresentable>)
```

"overload-style" -- this is already done elsewhere, e.g. for the constructor at https://docs.python.org/3/library/stdtypes.html#str -- but annoyingly verbose here...

```suggestion
.. class:: TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False, infer_variance=False)
           TypeVar(name, *constraints, default, bound=None, covariant=False, contravariant=False, infer_variance=False)
```

"man-page" style (does it make sense to mix "man-page" syntax with the Python syntax for defaults that the following parameters use?):

```suggestion
.. class:: TypeVar(name, *constraints[, default], bound=None, covariant=False, contravariant=False, infer_variance=False)
```

Author: Thanks, I like the `"<unrepresentable>"` variant best. The overload style is too verbose and mixing the bracket style and normal defaults looks weird.

> .. class:: TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False, infer_variance=False, default=<unrepresentable>)

Reviewer 2: May I suggest having two constructors? that way, the distinction between specifying a default and not specificying it would be perhaps clearer? 

```suggestion
.. class:: TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False, infer_variance=False)
		   TypeVar(name, *constraints, default, bound=None, covariant=False, contravariant=False, infer_variance=False)
```

Reviewer: Previously discussed at https://github.com/python/cpython/pull/116129#discussion_r1559924536

Reviewer 2: Ah thank you Alex (I should have read the previous comments more carefully!)

Author: I feel that would be less clear with such a long signature; it makes people search to spot the difference between the signatures.

         the constraints are evaluated only when the attribute is accessed, not when
         the type variable is created (see :ref:`lazy-evaluation`).

   .. attribute:: __default__

>       The default value of the type variable, if any. This attribute is ``None``

Reviewer 2: I think an example should be definitely helpful because people might really be confused between `TypeVar('T')` and `TypeVar('T', default=None)`.

Reviewer: Yeah, I'm somewhat concerned about the subtlety here. I just posted an alternative suggestion at https://discuss.python.org/t/pep-696-type-defaults-for-typevarlikes/22569/38, though it may be that it's too close to the beta deadline for us to consider making design changes to the PEP now.

Reviewer: (FWIW, I don't think that design question that I raised on Discourse should block the initial implementation here being merged.)

Reviewer 2: I don't think it should be a blocker, but I support your suggestion on discourse. 

Now, whatever is decided, the representation shouldn't change too many times afterwards. I think a similar issue (but I don't know whether it was formalized or not) is about the `<not given>` for enumerations, where we had some different representations across versions. This could definitely be some use cases for [PEP 661](https://peps.python.org/pep-0661/) (but I don't know if this one is actively needed). And depending on how you backports things, having a public value known to be the sentinel is better when you test representation outputs (the sentinel for enumerations is private).

Author: For BC reasons we probably can't afford to change this behavior later. I guess we can merge this version first and tweak the implementation during the beta phase, though.

Reviewer: > For BC reasons we probably can't afford to change this behavior later. I guess we can merge this version first and tweak the implementation during the beta phase, though.

yes, that was what I was thinking

Reviewer 2: Maybe it's obvious but what does "BC" mean in this context?

Reviewer: Backwards Compatibility

>       The default value of the type variable, if any. This attribute is ``None``

Reviewer: The docs here and for the other `__default__` attributes in `typing.rst` also need to be updated

         the constraints are evaluated only when the attribute is accessed, not when
         the type variable is created (see :ref:`lazy-evaluation`).

   .. attribute:: __default__

      The default value of the type variable, or :data:`typing.NoDefault` if it
      has no default.

      .. versionadded:: 3.13

   .. method:: has_default()

      Return whether or not the type variable has a default value. This is equivalent

>       to checking whether :attr:`__default__` is not equal to :data:`typing.NoDefault`,
>       except that it does not force evaluation of the

Reviewer: ```suggestion
      to checking whether :attr:`__default__` is not the :data:`typing.NoDefault`
      singleton, except that it does not force evaluation of the
```

      The name of the type variable tuple.

   .. attribute:: __default__

      The default value of the type variable tuple, or :data:`typing.NoDefault` if it
      has no default.

      .. versionadded:: 3.13

   .. method:: has_default()

      Return whether or not the type variable tuple has a default value. This is equivalent

>       to checking whether :attr:`__default__` is not equal to :data:`typing.NoDefault`,
>       except that it does not force evaluation of the

Reviewer: ```suggestion
      to checking whether :attr:`__default__` is not the :data:`typing.NoDefault`
      singleton, except that it does not force evaluation of the
```

      The name of the parameter specification.

   .. attribute:: __default__

      The default value of the parameter specification, or :data:`typing.NoDefault` if it
      has no default.

      .. versionadded:: 3.13

   .. method:: has_default()

      Return whether or not the parameter specification has a default value. This is equivalent

>       to checking whether :attr:`__default__` is not equal to :data:`typing.NoDefault`,
>       except that it does not force evaluation of the

Reviewer: ```suggestion
      to checking whether :attr:`__default__` is not the :data:`typing.NoDefault`
      singleton, except that it does not force evaluation of the
```

   .. versionadded:: 3.7.4

.. data:: NoDefault

>    A sentinel object used to represent the default value of certain parameters
>    used in this module.

Reviewer: ```suggestion
   A sentinel object used to indicate that a type parameter has no default
   value. For example:
   
   .. doctest::
   
      >>> T = TypeVar("T")
      >>> T.__default__ is typing.NoDefault
      True
      >>> S = TypeVar("S", default=None)
      >>> S.__default__ is None
      True
```

   .. versionadded:: 3.7.4

.. data:: NoDefault

   A sentinel object used to indicate that a type parameter has no default
   value. For example:

   .. doctest::

      >>> T = TypeVar("T")

>       >>> T.__default__ is typing.NoDefault
>       True
>       >>> S = TypeVar("S", default=None)
>       >>> S.__default__ is None
>       True

Reviewer: ```suggestion
      >>> T.has_default()
      False
      >>> T.__default__ is typing.NoDefault
      True
      >>> S = TypeVar("S", default=None)
      >>> S.has_default()
      True
      >>> S.__default__ is None
      True
```

Author: I thought of that but it felt out of place in the docs for `NoDefault`, since the cases you added don't use `NoDefault` at all.

Reviewer: Fair enough -- I thought it was quite nice to see together in one example how the two concepts interrelate, but I definitely don't feel strongly!
